tests/parser/WP_Parser_Node_Tests.php

```diff
@@ -0,0 +1,144 @@
+		$t_select  = new WP_Parser_Token( 100, 'SELECT' );
+		$t_comma   = new WP_Parser_Token( 200, ',' );
+		$t_plus    = new WP_Parser_Token( 300, '+' );
+		$t_one     = new WP_Parser_Token( 400, '1' );
+		$t_two_a   = new WP_Parser_Token( 400, '2' );
+		$t_two_b   = new WP_Parser_Token( 400, '2' );
+		$t_eof     = new WP_Parser_Token( 500, '' );
+
+		// Prepare a tree.
+		//
+		// A simplified testing tree for an input like "SELECT 1 + 2, 2".
+		//
+		// root
+		//   |- keyword
+		//   |    |- "SELECT"
+		//   |- expr [a]
+		//   |    |- "1"
+		//   |    |- "+"
+		//   |    |- expr [c]
+		//   |    |    |- "2" [b]
+		//   |- ","
+		//   |- expr [b]
+		//   |    |- "2" [a]
+		//   |- EOF
+		$root->append_child( $n_keyword );
+		$root->append_child( $n_expr_a );
+		$root->append_child( $t_comma );
+		$root->append_child( $n_expr_b );
+		$root->append_child( $t_eof );
+
+		$n_keyword->append_child( $t_select );
+		$n_expr_a->append_child( $t_one );
+		$n_expr_a->append_child( $t_plus );
+		$n_expr_a->append_child( $n_expr_c );
+		$n_expr_b->append_child( $t_two_a );
+		$n_expr_c->append_child( $t_two_b );
+
+		// Test "has" methods.
+		$this->assertTrue( $root->has_child() );
+		$this->assertTrue( $root->has_child_node() );
+		$this->assertTrue( $root->has_child_token() );
+
+		// Test single child methods.
+		$this->assertSame( $n_keyword, $root->get_child() );
+		$this->assertSame( $n_keyword, $root->get_child_node() );
+		$this->assertSame( $n_keyword, $root->get_child_node( 'keyword' ) );
+		$this->assertSame( $n_expr_a, $root->get_child_node( 'expr' ) );
+		$this->assertSame( $t_comma, $root->get_child_token() );
+		$this->assertSame( $t_comma, $root->get_child_token( 200 ) );
+		$this->assertNull( $root->get_child_token( 100 ) );
+
+		// Test multiple children methods.
+		$this->assertSame( array( $n_keyword, $n_expr_a, $t_comma, $n_expr_b, $t_eof ), $root->get_children() );
+		$this->assertSame( array( $n_keyword, $n_expr_a, $n_expr_b ), $root->get_child_nodes() );
+		$this->assertSame( array( $n_expr_a, $n_expr_b ), $root->get_child_nodes( 'expr' ) );
+		$this->assertSame( array(), $root->get_child_nodes( 'root' ) );
+		$this->assertSame( array( $t_comma, $t_eof ), $root->get_child_tokens() );
+		$this->assertSame( array( $t_comma ), $root->get_child_tokens( 200 ) );
+		$this->assertSame( array(), $root->get_child_tokens( 100 ) );
+
+		// Test single descendant methods.
+		// @TODO: Consider breadth-first search vs depth-first search.
+		$this->assertSame( $n_keyword, $root->get_descendant_node() );
+		$this->assertSame( $n_expr_a, $root->get_descendant_node( 'expr' ) );
+		$this->assertSame( null, $root->get_descendant_node( 'root' ) );
+		$this->assertSame( $t_comma, $root->get_descendant_token() );
+		$this->assertSame( $t_one, $root->get_descendant_token( 400 ) );
+		$this->assertSame( null, $root->get_descendant_token( 123 ) );
+
+		// Test multiple descendant methods.
+		// @TODO: Consider breadth-first search vs depth-first search.
```

Comment:
@adamziel I initially kept the logic from the first prototype (which was traversing descendants breadth-first), and added this note to look into it later, but now it seems to me it may be more useful and natural to do a dept-first search (which would return tokens in the order from the original query).

I'm still not 100% sure which one we need more — usually, I just want to know "is there this node or that token somewhere in this subtree", and in that case, both searches work. But sometimes it may be useful to check for a next token or more, for instance, and then `get_descendant_tokens()` would be more natural with a depth-first traversal. If really needed, we could also provide both. What do you think?